unit_tests/pipeline.cpp

```diff
@@ -0,0 +1,204 @@
+
+  void setup_two_stages_empty() {
+  }
+
+  void check_two_stages_empty() {
+    ASSERT_EQ(1, invocations_init); 
+    ASSERT_EQ(0, invocations_last); 
+  }
+
+  void setup_two_stages() {
+    out = 0;
+    counter = 2;
+  }
+
+  void check_two_stages() {
+    ASSERT_EQ(2, invocations_init); 
+    ASSERT_EQ(1, invocations_last); 
+    EXPECT_EQ(1, this->out);
+  }
+
+  void setup_three_stages() {
+    counter = 5;
+    out = 0;
+  }
+
+  void check_three_stages() {
+    ASSERT_EQ(5, invocations_init); 
+    ASSERT_EQ(4, invocations_last); 
+    ASSERT_EQ(4, invocations_intermediate);
+    EXPECT_EQ(20, this->out);
+  }
+
+  void setup_composed() {
+    counter = 5;
+    out = 0;
+
+  }
+
+  void check_composed() {
+    ASSERT_EQ(5, invocations_init); 
+    ASSERT_EQ(4, invocations_last); 
+    //ASSERT_EQ(4, invocations_intermediate);
+    EXPECT_EQ(40, this->out);
+  }
+
+
+};
+
+// Test for execution policies defined in supported_executions.h
+TYPED_TEST_CASE(pipeline_test, executions);
+
+TYPED_TEST(pipeline_test, static_two_stages_empty)
+{
+  this->setup_two_stages_empty();
+    grppi::pipeline( this->execution_,
+    [this]() { 
+        this->invocations_init++;
+        return optional<int>(); 
+    },
+    [this]( auto x ) {
+      this->invocations_last++;
+    }
+  );
+
+  this->check_two_stages_empty();
+}
+
+
+
+TYPED_TEST(pipeline_test, static_two_stages)
+{
+  this->setup_two_stages();
+    grppi::pipeline( this->execution_,
+    [this]() { 
+        this->invocations_init++;
+        this->counter--;
+        if(this->counter  == 0){
+          return optional<int>(); 
+        }else{
+          return optional<int>(this->counter);
+        }
+    },
+    [this]( auto x ) {
+      this->invocations_last++;
+      this->out += x;
+    }
+  );
+  this->check_two_stages();
+}
+
+
+
+TYPED_TEST(pipeline_test, static_three_stages)
+{
+  this->setup_three_stages();
+    grppi::pipeline( this->execution_,
+    [this]() { 
+        this->invocations_init++;
+        this->counter--;
+        if(this->counter  == 0){
+          return optional<int>(); 
+        }else{
+          return optional<int>(this->counter);
+        }
+    },
+    [this]( auto x ) {
+      this->invocations_intermediate++;
+      return x*2;
+    },
+    [this]( auto y ) {
+      this->invocations_last++;
+      this->out += y;
+    }
+  );
+  this->check_three_stages();
+}
+
+
+
+TYPED_TEST(pipeline_test, static_three_stages_composed)
+{
+  this->setup_composed();
+    grppi::pipeline( this->execution_,
+    [this]() { 
+        this->invocations_init++;
+        this->counter--;
+        std::vector<int> v(5);
+        std::iota(begin(v), end(v), 0);
+
+        if(this->counter  <= 0){
+          return optional< std::vector<int> >();
+        }else{
+          return optional<std::vector<int>>(v);
+        }
+    },
+    grppi::farm(this->execution_,
+        [this](std::vector<int> v) {
+          //this->invocations_intermediate++;
```

Comment:
This needs to be debugged and solved.

A core is generated with native and TBB.